### DISCUSSION TITLE
GitHub Advisory Database searching

### DISCUSSION THREAD

**Discussion by nikoladsp** [community/community#40335] [Apps, API and Webhooks] [answered]
### Select Topic Area

Question

### Body

Hi,
I would like to know how to search GitHub Advisory Database  programatically - either using REST or GraphQL. For example, I would like to get response similar to when I call: https://github.com/advisories?query=keycloak+type%3Areviewed

I tried to figure out myself, but did not get far. I can use something like:

```
{
  securityAdvisories(first: 100) {
    totalCount
    pageInfo {
      endCursor
      hasNextPage
    }
    nodes {
      summary
      ghsaId
      severity
    }
  }
}
```

But I would have to call this 100 times and filter summary each time, which lasts way too long. 

Many thanks in advance

**Comment by LangLangBart** [comment, marked as answer]
> But I would have to call this 100 times and filter summary each time, which lasts way too long.

Using [GitHub's cli](https://github.com/cli/cli) tool took ~1min.
```zsh
❯ QUERY='query ($endCursor: String) {
  securityAdvisories(first: 100, after: $endCursor) {
    pageInfo {
      endCursor
      hasNextPage
    }
    nodes {
      summary
      ghsaId
      severity
    }
    totalCount
  }
}'

❯ time gh api graphql --silent --paginate --raw-field query=$QUERY
user=0.36s system=0.11s cpu=0% total=1:01.69
```

`gh` has `jq` built in to narrow the query to the data we want.
```zsh
❯ gh api graphql --paginate --raw-field query=$QUERY --jq '.data.securityAdvisories.nodes[] | select(.summary | test("Keycloak";"i"))'
# {"ghsaId":"GHSA-755v-r4x4-qf7m","severity":"MODERATE","summary":"Stored Cross-Site Scripting (XSS) in Keycloak via groups dropdown"}
# {"ghsaId":"GHSA-qpq9-jpv4-6gwr","severity":"HIGH","summary":"Keycloak allows anyone to register new security device or key for any user by using WebAuthn password-less login flow"}
# {"ghsaId":"GHSA-237q-6hjp-pchq","severity":"MODERATE","summary":"JBoss KeyCloak is vulnerable to soft token deletion via CSRF"}
# {"ghsaId":"GHSA-2mq8-99q7-55wx","severity":"HIGH","summary":"Code injection in keycloak"}
...
```

#### possible improvements
- adding the cache flag could also help, for example `--cache 5m`
- to improve performance, you may add more arguments to limit the number of results, for example: `updatedSince: "2022-10-01T00:00:00"` or
- try to limit it to a certain ecosystem, for example with `securityVulnerabilities` `ecosystem: MAVEN`
```zsh
❯ QUERY='query ($endCursor: String) {
  securityVulnerabilities(first: 100, after: $endCursor, ecosystem: MAVEN) {
    totalCount
    pageInfo {
      endCursor
      hasNextPage
    }
    nodes {
      advisory {
        summary
        ghsaId
        severity
      }
    }
  }
}'
❯ time gh api graphql --silent --paginate --raw-field query=$QUERY
user=0.22s system=0.06s cpu=1% total=20.545

gh api graphql --paginate --raw-field query=$QUERY --jq '.data.securityVulnerabilities.nodes[].advisory | select(.summary | test("Keycloak";"i"))'
# {"ghsaId":"GHSA-755v-r4x4-qf7m","severity":"MODERATE","summary":"Stored Cross-Site Scripting (XSS) in Keycloak via groups dropdown"}
# {"ghsaId":"GHSA-237q-6hjp-pchq","severity":"MODERATE","summary":"JBoss KeyCloak is vulnerable to soft token deletion via CSRF"}
# {"ghsaId":"GHSA-jh7q-5mwf-qvhw","severity":"MODERATE","summary":"Keycloak vulnerable to Server-Side Request Forgery"}
...
```

---

what is missing is a `query` argument to make better use of the `Advisory` qualifiers (see [Searching the GitHub Advisory Database](https://docs.github.com/en/code-security/security-advisories/global-security-advisories/browsing-security-advisories-in-the-github-advisory-database#searching-the-github-advisory-database))

<img src="https://user-images.githubusercontent.com/92653266/204664099-6f975c6c-f479-44e7-8f8e-f81f8f283840.png" width="300">

The `search` has such a `query`, but it does not allow to search `Advisory` types, only Issue/PR, Discussion, User and Repo are supported.
<img src="https://user-images.githubusercontent.com/92653266/204666226-e91e5026-ec89-4c38-9761-d4d143dff0b5.png" width="400">

**Comment by nikoladsp** [comment]
I've got similar results (approx 75 secs, but searching summary for each returned record) - not sure how https://github.com/advisories is so fast and it searches not only summary but sources and references field

I will explore these suggestions - many thanks; you opened few possibilities!
